<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Définir-pour-accélerer" data-toc-modified-id="Définir-pour-accélerer-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Définir pour accélerer</a></span></li></ul></li><li><span><a href="#Développer-des-fonctions" data-toc-modified-id="Développer-des-fonctions-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Développer des fonctions</a></span><ul class="toc-item"><li><span><a href="#Gestion-de-erreurs," data-toc-modified-id="Gestion-de-erreurs,-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Gestion de erreurs,</a></span></li><li><span><a href="#Résultat-en-plusieurs-éléments-(list)" data-toc-modified-id="Résultat-en-plusieurs-éléments-(list)-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Résultat en plusieurs éléments (<code>list</code>)</a></span></li><li><span><a href="#inspecter-la-fonction:" data-toc-modified-id="inspecter-la-fonction:-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>inspecter la fonction:</a></span><ul class="toc-item"><li><span><a href="#le-code" data-toc-modified-id="le-code-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>le code</a></span></li><li><span><a href="#les-arguments-de-la-fonction" data-toc-modified-id="les-arguments-de-la-fonction-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>les arguments de la fonction</a></span></li><li><span><a href="#Les-arguments-par-défaut" data-toc-modified-id="Les-arguments-par-défaut-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Les arguments par défaut</a></span></li></ul></li><li><span><a href="#Best-practice" data-toc-modified-id="Best-practice-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Best practice</a></span></li></ul></li><li><span><a href="#Sauvegegarde-d'objets" data-toc-modified-id="Sauvegegarde-d'objets-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Sauvegegarde d'objets</a></span></li></ul></div>

On se rappelle qu'au dernier cours, on avait fait `moyennes<-NULL`

In [43]:
moyennes<-NULL
a<-system.time(
for (i in 1:20000){
    mu<-mean(rnorm(100))
    moyennes<-rbind(moyennes, mu)
}
)
a

   user  system elapsed 
  3.953   0.107   4.063 

## Définir pour accélerer

Au lieu de créer un objet `NULL` de taille et de type indéfini et d'y insérer le calcul via `rbind`, on peut créer un vecteur vide de même taille que le nombre de valeurs du résultat produit.


In [46]:
moyennes<-numeric(length = 20000)
b<-system.time(
for (i in 1:20000){
    mu<-mean(rnorm(100))
    moyennes[i] <-mu
}
)
b

   user  system elapsed 
  0.413   0.002   0.415 

Ce qu'on vient de faire est génial! au lieu d'ajouter le résultat calculé à l'objet vide à chaque itération, nous avons créé un vecteur en définit sa taille, et nous y avons assigné plutôt le résultat à l'index courant.

Comparons les temps écoulés entre les deux manières de faire:


In [45]:
round(a[3]/b[3],2)

elapsed 
   9.32

Donc la première façon de faire et 9.32 fois plus lente que la deuxième!

# Développer des fonctions

Dans les chapitres précédents, nous avons développé plusieurs exemples de code. Par exemple, nous avons calculé la valeur présente d'une série de paiements. Dans cette section, nous verrons qu'il est possible de faire ces mêmes calculs en se servant de nos propres fonctions. Creer des fonctions nous permet d'éviter de réécrire le code qui fait le même calcule à chaque fois.


Syntaxe:
```{R}
NonDeFonction <- function(arg1, arg2, ...) {
    codeAexécuter
    return(resultatSouhaité)
}
```

prenons l'exemple classique où l'on crée une fonction qui additionne deux éléments;

In [47]:
somme<-function(x,y){
    return(x+y)
}

Appliquons cette fonction maintenant:

In [48]:
somme(2,3)

[1] 5

## Gestion de erreurs,

On peut éviter des erreurs d'exécution par le moyen de conditions. Toutesfois, cette solution nécessite une connaissance des erreurs qui peuvent survenir. Dans notre exemple d'adition de deux éléments, on peut demander au programme de vérifier si les éléments à additionner sont d'abord de type numérique;

In [49]:
somme<-function(x,y){
    if(is.numeric(x) & is.numeric(y)){
        return(x+y)
    }else {
        print("veuillez vérifier vos arguments")
    }      
}

In [50]:
somme(2,2)

[1] 4

Toutefois, lorsqu'on essaie ceci:

In [51]:
somme(2, "a")

[1] "veuillez vérifier vos arguments"


Dans l'exemple précédent, nous avons eu comme résultat la somme de deux éléments, il est possible de faire en sorte que le résultat de la fonction nous retourne plusieurs éléments. Par exemple, dans la fonction `operation` suivante, elle nous retourne trois opérations sur les deux arguments de la fonction; la somme, la différence,la multiplication , ainsi que la division des deux éléments;



In [53]:
operation<-function(x,y){
    if(is.numeric(x) & is.numeric(y)){
        return(c(x+y, x-y, x*y, x/y))
    }else {
        print("veuillez vérifier vos arguments")
    }        
}

In [54]:
operation(2,3)

[1]  5.0000000 -1.0000000  6.0000000  0.6666667

## Résultat en plusieurs éléments (`list`)

Au lieu d'utiliser `return`, on peut utiliser `list` afin d'avoir une liste d'éléments dans le résultat. L'avantage d'avoir une liste est bien sûr la manipulation d'élément de cette dernière.


In [55]:
operation_l<-function(x,y){
    if(is.numeric(x) & is.numeric(y)){
        list(x+y, x-y, x*y, x/y)
    }else {
        print("veuillez vérifier vos arguments")
    }        
}

In [58]:
resultat_sousForme_liste<-operation_l(2,3)
resultat_sousForme_liste

[[1]]
[1] 5

[[2]]
[1] -1

[[3]]
[1] 6

[[4]]
[1] 0.6666667

Ainsi, on peut extraire **un seul** (ou plusieurs) élément de notre liste résultante;

In [59]:
resultat_sousForme_liste[2]

[[1]]
[1] -1

Au lieu d'insérer deux éléments comme argument de notre fonction `operation_l`, insérons deux vecteurs et regardons la différence entre le résultat de la fonction `operation_l` et la fonction `operation`

In [60]:
xvec<-c(1:5)
yvec<-c(6:10)
operation_l(xvec, yvec)

[[1]]
[1]  7  9 11 13 15

[[2]]
[1] -5 -5 -5 -5 -5

[[3]]
[1]  6 14 24 36 50

[[4]]
[1] 0.1666667 0.2857143 0.3750000 0.4444444 0.5000000

Maintenant, appliquons la fonction `operation` sur les mêmes vecteurs et inspectons à nouveau le résultat:

In [61]:
operation(xvec, yvec)

[1]  7.0000000  9.0000000 11.0000000 13.0000000 15.0000000 -5.0000000
 [7] -5.0000000 -5.0000000 -5.0000000 -5.0000000  6.0000000 14.0000000
[13] 24.0000000 36.0000000 50.0000000  0.1666667  0.2857143  0.3750000
[19]  0.4444444  0.5000000

Si l'on voulait par exemple extraire le deuxième élément de la liste, qui est le résultat de la soustraction dans notre exemple, il est plus difficile de le faire avec la fonction `operation`.

In [62]:
operation_l(xvec, yvec)[2]

[[1]]
[1] -5 -5 -5 -5 -5

au lieu de:

In [63]:
operation(xvec, yvec)[6:10]

[1] -5 -5 -5 -5 -5

Où il est nécessaire de connaitre les index du seul vecteur résultant.

## inspecter la fonction:

On peut inspecter le code de notre fonction en inscrivant le nom de cette dernière. Surtout, lorsque le code de la fonction est perdu quelque part dans un autre code 100 lignes plutôt.



### le code

In [64]:
operation_l

function(x,y){
    if(is.numeric(x) & is.numeric(y)){
        list(x+y, x-y, x*y, x/y)
    }else {
        print("veuillez vérifier vos arguments")
    }        
}

### les arguments de la fonction

On peut obtenir les arguments d'une fonction avec l'instruction suivante;


In [65]:
formals(operation_l)

$x


$y



### Les arguments par défaut

Il convient souvent de donner des arguments par défaut à notre fonction créée. Dans notre exemple `operation_l`, on pourrait données les valeurs `(x,y)={1,1}` afin d'éviter des division par `0` par exemple;



In [66]:
operation_l<-function(x=1,y=1){
    if(is.numeric(x) & is.numeric(y)){
        list(x+y, x-y, x*y, x/y)
    }else {
        print("veuillez vérifier vos arguments")
    }        
}

Lorsqu'on écrit la fonction sans aucun argument;

In [67]:
operation_l()

[[1]]
[1] 2

[[2]]
[1] 0

[[3]]
[1] 1

[[4]]
[1] 1

On obtient le résultat de nos valeurs par défaut

## Best practice

Il est très conseillé de bien documenter les fonctions que vous créez. Cela vous aide lorsque vous revisez ou corrigez votre code. Sans oublier que la bonne documentation aide la personne qui relie votre code à bien comprendre ce que vous voulez avoir comme résultat. Que ce soit dans un contexte académique (travaux, examen....etc.) Et surtout dans le contexte professionnel, où les codes sont plus longs et plus lourds à lire lorsque la documentation est quasi absente.


In [68]:
operation_l<-function(x=1,y=1){
  # Cette fonction calcule l'adition, soustraction, multiplication
  # ainsi que la division de deux vecteurs
  # Args:
  #   x: La valeur du premier vecteur, la valeur par défaut étant =1
  #   y: La valeur du deuxième vecteur, la valeur par défaut étant =1
  #
  # retourne:
  #   une liste du résultat des opérations ....      

    if(is.numeric(x) & is.numeric(y)){
        list(x+y, x-y, x*y, x/y)
    }
    # Gestion d'erreurs
    else {
        print("veuillez vérifier vos arguments")
    }        
} 

# Sauvegegarde d'objets

Il possible de sauvegarder tout objet sous format `.RData`. Il convient de faire cette sauvegarde lorsque nous ne voulons pas garder nos objets en mémoire. Surtout, lorsque nous travaillons avec des données importantes en volume.


In [69]:
x<-matrix(rnorm(100), ncol=10)

In [70]:
save(x, file = "x.RData")

Maintenant, on peut recharger ces mêmes données. pour être sûr que nous n'avons pas chargé les données qui étaient en mémoire, supprimons tout ce qui se retrouve dans l'environnement:

In [71]:
rm(list=ls())

In [72]:
load("x.RData")

In [73]:
x

[,1]        [,2]        [,3]        [,4]        [,5]       [,6]       
 [1,]  1.18939619  0.01057899 -0.35654252  1.65102435 -1.3475614 -0.70525999
 [2,] -0.06534286  1.42791869  0.87576387 -1.04904302 -1.2970706 -0.54152388
 [3,] -1.10988733  0.40814513 -0.98567474  0.07558152 -0.1571415  0.09037718
 [4,]  0.67465672  0.84057229 -0.03364994  1.78939415  0.9036412  0.13285054
 [5,]  3.34160414 -0.47974054 -0.67199908  0.85596379  0.1871276 -0.34076899
 [6,] -1.79255793 -0.99787722  2.70522950  1.60416877 -1.1513655  0.54021006
 [7,] -0.16944079  1.25588922 -0.12954850  0.37858396 -0.3957809  0.80730431
 [8,]  0.90407239 -0.27608180 -0.08992006 -0.23410372 -1.6570973 -1.59602886
 [9,]  0.65195276 -0.54867528  1.46378626  0.59448827 -0.9709800 -0.93753977
[10,]  1.74168331  0.18004910  0.11510721 -0.59757236 -0.2167422 -0.01264427
      [,7]        [,8]       [,9]       [,10]      
 [1,]  0.12251689 -1.9915918 -0.3322830  1.02455290
 [2,]  1.24576523  0.6561742 -0.4844493 -0.39839616
 [3,] -0.56200895 -0.4960402 -1.2550586 -0.07348674
 [4,]  0.72001113  1.5614279 -1.2266363  1.71525509
 [5,]  0.05979965 -0.7835429  1.0479630  0.08575501
 [6,] -0.65092374  0.7824136  1.4823115  0.63362158
 [7,]  0.29098695 -1.0314703 -2.5441580 -1.20221796
 [8,]  0.62270556  0.4811070 -0.2526573  1.77087313
 [9,] -0.05390880  0.7824050 -0.1869861 -1.14069893
[10,]  1.09123924 -0.8519077 -1.7291689  0.05694140

Il aussi possible de sauvegarder les données (les résultats par exemple) dans des fichiers `.csv` qu'on pourrait redistribuer par la suite dans le contexte de production de rapport ou de partage de résultats.

Nous utilisons alors la fonction `write.csv`;


In [74]:
write.csv(x, "/Users/nour/Downloads/données_x.csv")

Remarquez qu'à chaque fois nous réexécutons ce code, nous supprimons l'ancien fichier portant le même nom.

On peut aussi lire ce même fichier avec la fonction 


In [75]:
oo<-read.csv("/Users/nour/Downloads/données_x.csv")

In [76]:
head(oo)

X V1          V2          V3          V4          V5         V6         
1 1  1.18939619  0.01057899 -0.35654252  1.65102435 -1.3475614 -0.70525999
2 2 -0.06534286  1.42791869  0.87576387 -1.04904302 -1.2970706 -0.54152388
3 3 -1.10988733  0.40814513 -0.98567474  0.07558152 -0.1571415  0.09037718
4 4  0.67465672  0.84057229 -0.03364994  1.78939415  0.9036412  0.13285054
5 5  3.34160414 -0.47974054 -0.67199908  0.85596379  0.1871276 -0.34076899
6 6 -1.79255793 -0.99787722  2.70522950  1.60416877 -1.1513655  0.54021006
  V7          V8         V9         V10        
1  0.12251689 -1.9915918 -0.3322830  1.02455290
2  1.24576523  0.6561742 -0.4844493 -0.39839616
3 -0.56200895 -0.4960402 -1.2550586 -0.07348674
4  0.72001113  1.5614279 -1.2266363  1.71525509
5  0.05979965 -0.7835429  1.0479630  0.08575501
6 -0.65092374  0.7824136  1.4823115  0.63362158